pkg/argo_client/manifests.go

```diff
@@ -1,116 +1,470 @@
+func copyFile(srcpath, dstpath string) error {
+	dstdir := filepath.Dir(dstpath)
+	if err := os.MkdirAll(dstdir, 0o777); err != nil {
+		return errors.Wrap(err, "failed to make directories")
+	}
+
+	r, err := os.Open(srcpath)
+	if err != nil {
+		return err
+	}
+	defer r.Close() // ignore error: file was opened read-only.
+
+	w, err := os.Create(dstpath)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		// Report the error, if any, from Close, but do so
+		// only if there isn't already an outgoing error.
+		if c := w.Close(); err == nil {
+			err = c
+		}
+	}()
+
+	_, err = io.Copy(w, r)
+	return err
+}
+
+func packageApp(ctx context.Context, source v1alpha1.ApplicationSource, refs []v1alpha1.ApplicationSource, repo *git.Repo, getRepo getRepo) (string, error) {
+	tempDir, err := os.MkdirTemp("", "package-*")
+	if err != nil {
+		return "", errors.Wrap(err, "failed to make temp dir")
+	}
+
+	tempAppDir := filepath.Join(tempDir, source.Path)
+	appPath := filepath.Join(repo.Directory, source.Path)
+
+	// copy app files to the temp dir
+	if err = filepath.Walk(appPath, func(path string, info fs.FileInfo, err error) error {
 		if err != nil {
-			log.Warn().Err(err).Msg("Failed to format manifest")
-			continue
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		relPath, err := filepath.Rel(appPath, path)
+		if err != nil {
+			return errors.Wrapf(err, "failed to calculate rel between %q and %q", appPath, path)
+		}
+		src := path
+		dst := filepath.Join(tempAppDir, relPath)
+		if err := copyFile(src, dst); err != nil {
+			return errors.Wrapf(err, "failed to %s => %s", src, dst)
 		}
-		manifests = append(manifests, fmt.Sprintf("---\n%s", string(ret)))
+		return nil
+	}); err != nil {
+		return "", errors.Wrap(err, "failed to copy files")
 	}
-	return manifests
+
+	if source.Helm != nil {
+		refsByName := make(map[string]v1alpha1.ApplicationSource)
+		for _, ref := range refs {
+			refsByName[ref.Ref] = ref
+		}
+
+		for index, valueFile := range source.Helm.ValueFiles {
+			if strings.HasPrefix(valueFile, "$") {
+				relPath, err := processValueReference(ctx, source, valueFile, refsByName, repo, getRepo, tempDir, tempAppDir)
+				if err != nil {
+					return "", err
+				}
+
+				source.Helm.ValueFiles[index] = relPath
+				continue
+			}
+
+			if strings.Contains(valueFile, "://") {
+				// todo: is there anything to do here?
+				continue
+			}
+
+			relPath, err := filepath.Rel(source.Path, valueFile)
+			if err != nil {
+				return "", errors.Wrap(err, "failed to calculate relative path")
+			}
+
+			if !strings.HasPrefix(relPath, "../") {
+				continue // this values file is already copied
+			}
+
+			src := filepath.Join(appPath, valueFile)
+			dst := filepath.Join(tempAppDir, valueFile)
+			if err = copyFile(src, dst); err != nil {
+				return "", errors.Wrapf(err, "failed to copy file: %q", valueFile)
+			}
+		}
+	}
+
+	return tempDir, nil
+}
+
+func processValueReference(
```

Comment:
can we also unitest this? we can mock the `getRepo` func, no?

Reply:
This should be tested as part of packageApp, and that test is gonna suck enough that I'm hoping it takes care of both 🤞 coverage should show us if that's true.